zipkin-storage/cassandra-v1/src/main/java/zipkin2/storage/cassandra/v1/IndexTraceId.java

```diff
@@ -0,0 +1,243 @@
+ * the License.
+ */
+package zipkin2.storage.cassandra.v1;
+
+import com.datastax.driver.core.BoundStatement;
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.ResultSetFuture;
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.querybuilder.Insert;
+import com.datastax.driver.core.querybuilder.QueryBuilder;
+import com.google.auto.value.AutoValue;
+import com.google.common.cache.CacheBuilder;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import zipkin2.storage.QueryRequest;
+import zipkin2.storage.cassandra.internal.call.DeduplicatingVoidCallFactory;
+import zipkin2.storage.cassandra.internal.call.ResultSetFutureCall;
+
+/**
+ * Inserts index rows into a Cassandra table. This skips entries that don't improve results based on
+ * {@link QueryRequest#endTs()} and {@link QueryRequest#lookback()}. For example, it doesn't insert
+ * rows that only vary on timestamp and exist between timestamps of existing rows.
+ */
+final class IndexTraceId extends ResultSetFutureCall<Void> {
+  static final Logger LOG = LoggerFactory.getLogger(IndexTraceId.class);
+
+  @AutoValue
+  abstract static class Input {
+    static Input create(String partitionKey, long timestamp, long traceId) {
+      return new AutoValue_IndexTraceId_Input(partitionKey, timestamp, traceId);
+    }
+
+    abstract String partitionKey(); // ends up as a partition key, ignoring bucketing
+
+    abstract long ts(); // microseconds at millis precision
+
+    abstract long trace_id(); // clustering key
+  }
+
+  /** Deduplicates redundant entries upon {@link #iterator()} */
+  interface Indexer extends Iterable<IndexTraceId.Input> {
+    Indexer NOOP = new Indexer() {
+      @Override public void add(Input input) {
+      }
+
+      @Override public Iterator<Input> iterator() {
+        return Collections.emptyIterator();
+      }
+    };
+
+    void add(IndexTraceId.Input input);
+  }
+
+  static abstract class Factory extends DeduplicatingVoidCallFactory<Input> {
+    final Session session;
+    // Shared across all threads as updates can come from any thread.
+    final Map<Entry<String, Long>, Pair> sharedState;
+    final String table;
+    final int bucketCount;
+    final PreparedStatement preparedStatement;
+    final TimestampCodec timestampCodec;
+
+    Factory(CassandraStorage storage, String table, int indexTtl) {
+      super(TimeUnit.SECONDS.toMillis(storage.indexCacheTtl), storage.indexCacheTtl);
+      session = storage.session();
+      // TODO: this state should be merged with the DelayLimiter setup in the super class
```

Comment:
in reality this might not be possible to merge, as the data is used for different purposes. For example, here it is used to determine if an existing range will be extended.. not if an exact call was made prior.

As the main next goal is getting rid of guava, the follow-up PR may not actually merge these features into DelayLimiter, rather make a similar bounded expire-after-write map.